### ISSUE TITLE
Add static styles to non-shadow elements

### ISSUE THREAD

**Issue by davidbwaters** [lit/lit#1977]
Is there a reason why styling can't be applied to light-dom elements?  I understand global styles apply to these, but writing them in the same place as part of the component is a huge advantage. 

This would make elements behave consistently when changing the render root without having to refactor.  This would make the convention of writing styles there consistent across all components.  

Styling could be appended to the document head, or another method.  

A common scenario where shadow-dom doesn't play well is using libraries that select elements from the document root, which includes a ton of frontend libs.

**Comment by justinfagnani** [comment]
Yes, the reason is that we don't have a good scope to add the styles to. We can't write to the parent's scope, as that breaks upper-bound encapsulation and may break the parent's styles. We can't apply to document because the document might not container the element's DOM if there's an ancestor with a shadow root. If the scope is not the element's shadow root, selectors like `:host` and `::slotted` will select the wrong elements.

If you want to apply styles to the document I'd suggest just doing that manually in the module:

```ts
const style = document.createElement('style');
style.innerHTML = `
  my-element { ... }
  my-element > .child1 { ... }
`;

export class MyElement extends LitElement {
  createRenderRoot() { return this; }
}
```

**Comment by davidbwaters** [comment]
I understand `:host` and `:slotted` wouldn't work, but I think understanding usage of these implies understanding that these only apply to the shadow root.  I understand not wanting to break encapsulation and that makes sense, but I would still prefer some sort or default behavior that would make these work.  They could be applied in a style tag before `render()` content. Understanding the difference between shadow and light dom, it's understood these styles could potentially conflict with existing styles.

**Comment by sorvell** [comment]
We really want to encourage users to use Shadow DOM since it's the right way to do DOM and style encapsulation. That said, a lot of users have these styling concerns. While it's possible to do something like Justin suggested, the next thing you'll run into is that `<slot>`'s don't work without Shadow DOM. Addressing that is outside the scope of something we want to support, but it can be done by using the Shadow DOM polyfill and forcing it to be on. 

If you don't use slots, something like what you're asking for is more possible but you definitely have to be really careful not to stomp on styling used elsewhere in the document. In particular, it's tricky to enforce a lower bound encapsulation using normal-ish css and typically this requires a system that everyone on the page must opt in.

With all these caveats in mind, [here is an example](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudCwgQ1NTUmVzdWx0R3JvdXAsIENTU1Jlc3VsdCwgQ1NTUmVzdWx0T3JOYXRpdmUsIHVuc2FmZUNTU30gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuY29uc3QgdGFnID0gJ3NpbXBsZS1ncmVldGluZyc7XG5cbi8vIFNvbWUgcmVhbGx5IGJhc2ljIHRyYW5zZm9ybWF0aW9uIG9mIHNoYWRvdyB0byBsaWdodCBET00gc3R5bGVzIHdpdGggbWluaW1hbCBlbmNhcHN1bGF0aW9uLlxuY29uc3Qgc2VsID0gKGU6IHR5cGVvZiBMaXRFbGVtZW50ICYge3RhZzogc3RyaW5nLCBkaXNhYmxlU2hhZG93Um9vdDogYm9vbGVhbn0sIHNlbGVjdG9yOiBzdHJpbmcpID0-IHtcbiAgaWYgKGUuZGlzYWJsZVNoYWRvd1Jvb3QpIHtcbiAgICBpZiAoIXNlbGVjdG9yLmluY2x1ZGVzKCc6aG9zdCcpKSB7XG4gICAgICBzZWxlY3RvciA9IGA6aG9zdCAke3NlbGVjdG9yfWA7XG4gICAgfVxuICAgIHNlbGVjdG9yID0gc2VsZWN0b3IucmVwbGFjZSgvXFw6aG9zdC9nLCBlLnRhZyk7XG4gIH1cbiAgcmV0dXJuIHVuc2FmZUNTUyhzZWxlY3Rvcik7XG59XG5cbkBjdXN0b21FbGVtZW50KHRhZylcbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBcbiAgc3RhdGljIHRhZyA9IHRhZztcbiAgXG4gIHN0YXRpYyBkaXNhYmxlU2hhZG93Um9vdCA9IHRydWU7XG4gIFxuICBzdGF0aWMgZ2V0IHN0eWxlcygpIHsgcmV0dXJuIGNzc2BcbiAgICAgICR7c2VsKHRoaXMsICc6aG9zdCcpfSB7XG4gICAgICAgIGRpc3BsYXk6IGlubGluZS1ibG9jaztcbiAgICAgICAgYm9yZGVyOiAycHggZGFzaGVkIG9yYW5nZTtcbiAgICAgICAgcGFkZGluZzogNHB4O1xuICAgICAgfVxuICAgICAgXG4gICAgICAke3NlbCh0aGlzLCAnZGl2Jyl9IHsgY29sb3I6IGJsdWUgfVxuICBgO1xuIH1cblxuICBAcHJvcGVydHkoKVxuICBuYW1lID0gJ1NvbWVib2R5JztcblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPGRpdj5IZWxsbywgJHt0aGlzLm5hbWV9ITwvZGl2PmA7XG4gIH1cbiAgXG4gIHByb3RlY3RlZCBzdGF0aWMgZmluYWxpemVTdHlsZXMoXG4gICAgc3R5bGVzPzogQ1NTUmVzdWx0R3JvdXBcbiAgKTogQ1NTUmVzdWx0T3JOYXRpdmVbXSB7XG4gICAgbGV0IGVsZW1lbnRTdHlsZXMgPSBzdXBlci5maW5hbGl6ZVN0eWxlcyhzdHlsZXMpO1xuICAgIGNvbnN0IHN0eWxlUm9vdCA9IGRvY3VtZW50LmhlYWQ7XG4gICAgaWYgKHRoaXMuZGlzYWJsZVNoYWRvd1Jvb3QpIHtcbiAgICAgIC8vIFdBUk5JTkc6IFRoaXMgYnJlYWsgY29tcG9uZW50IGVuY2Fwc3VsYXRpb24gYW5kIGFwcGxpZXMgc3R5bGVzIHRvIHRoZSBkb2N1bWVudC4gXG4gICAgICAvLyBUaGVzZSBzdHlsZXMgc2hvdWxkIGJlIG1hbnVhbGx5IHNjb3BlZC5cbiAgICAgIGVsZW1lbnRTdHlsZXMuZm9yRWFjaCgoczogQ1NTUmVzdWx0KSA9PiB7XG4gICAgICAgIGNvbnN0IHN0eWxlID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnc3R5bGUnKTtcbiAgICAgICAgc3R5bGUudGV4dENvbnRlbnQgPSBzLmNzc1RleHQ7XG4gICAgICAgIHN0eWxlUm9vdC5hcHBlbmRDaGlsZChzdHlsZSk7XG4gICAgICB9KTtcbiAgICAgIGVsZW1lbnRTdHlsZXMgPSBbXTtcbiAgICB9XG4gICAgcmV0dXJuIGVsZW1lbnRTdHlsZXM7XG4gIH1cbiAgXG4gIGNyZWF0ZVJlbmRlclJvb3QoKSB7XG4gICAgcmV0dXJuICh0aGlzLmNvbnN0cnVjdG9yIGFzIGFueSkuZGlzYWJsZVNoYWRvd1Jvb3QgPyB0aGlzIDogc3VwZXIuY3JlYXRlUmVuZGVyUm9vdCgpO1xuICB9XG59XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIj48L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9XQ) that shows how the `createRenderRoot` and `finalizeStyles` API can be used to customize the styling behavior. Hope that helps.

**Comment by davidbwaters** [comment]
Thanks, like I mentioned in the OP, the main use-case for using the light DOM is lib compatibility.

But I do think it should be considered how opinionated Lit wants to be about Shadow DOM (or anything, really).  I've heard arguments making the case that encapsulation is overrated, which I don't necessarily agree with, but I can understand people favoring architecting code to use global scope responsibly to your advantage.  Plus, not everyone is working on a project at scale.  

I never really got off of the BEM bandwagon, so style encapsulation has never really been that big of a deal to me because I was pretty much building it in myself with a naming convention.  There are times when I think asset loading would be more efficient being able to global utility classes outside of slots.

I think Lit wants to tread the line of enforcing certain good practices, but remain general purpose, like React, etc.  I think there are plenty of scenarios or workflows where Shadow could be considered overkill.  Plus, I think web component adoption (aside from the API issues and support) might have been slower because people equated custom elements with Shadow-DOM, when some people just want to write reusable chunks of code.  

Also, I think React's success also came from areas where it is unopinionated and creative use cases emerged that may not have been considered during inception, like server components and primitives, etc.

**Comment by davidbwaters** [comment]
Also, if you consider the POV of a BEM-user, like myself, and I think this is a good example, scoped styles offer zero advantage because you are namespaced by default, but I'd still like certain structural styles to be able to be shared in the component definition in a consistent way.  

That leaves slots as the coolest part of shadow-DOM, but they require a higher-level of understanding of light and shadow DOM boundaries and I could see this being a source of confusion for people that are learning.  

I'm only now getting completely comfortable with how slotted content behaves differently than anything else and I remember staring at my inspector trying to figure out WTH is going on for the first time, with content 'reflecting' inside.  I think being less opinionated in places where best practices can be debated would be a good rule of thumb.  

Not having an all-or-nothing approach with Shadow DOM would also let people adopt Lit progressively and start using Shadow when they are comfortable, which is another thing that I think React benefited from, as far as adoption goes.  You could get started just returning nodes and use building blocks and the rest you could pick up as you saw fit.

Especially now that a 'back to basics' approach has everyone considering 'how much is too much?' and your own demos went from an html file with a single app element to a static site generator, it's only fair to apply that same thinking to Shadow.

**Comment by sorvell** [comment]
This isn't something we want to directly support in the core library, but you should feel free to make your own base class that customizes the behavior here. Please open a new issue if you need any guidance on how/what to do for this use case. Thanks!

**Comment by justinfagnani** [comment]
I'm going to reopen this as a documentation issue, because I think there might be a way to self-service by also overriding adoptStyles.

**Comment by Himself65** [comment]
Lit should expose a helper function or class that automatically writes style to the top document if it's not a shadow dom.

At least, some CSS is better than no CSS at all.

**Comment by mucsi96** [comment]
```js
  createRenderRoot() {
    const style = document.createElement("style");
    style.innerHTML = this.constructor.elementStyles;
    document.head.append(style);
    return this;
  }
```

solved the problem for me. But fully agree with @Himself65 and @justinfagnani. Would be really helpfull if we could have something like `adoptStyles` in [lit](https://github.com/lit/lit/tree/main)/[packages](https://github.com/lit/lit/tree/main/packages)/[reactive-element](https://github.com/lit/lit/tree/main/packages/reactive-element)/[src](https://github.com/lit/lit/tree/main/packages/reactive-element/src)
/css-tag.ts which supports non-shadow elements.

**Comment by AliMD** [comment]
I used this function to handle it (without any scope)

```ts
function flatCssText(styles?: CSSResultGroup): string {
  if (styles === undefined) return '';
  if ('cssText' in styles) return styles.cssText.trim();
  if (Array.isArray(styles)) {
    return styles.map(style => ('cssText' in style) ? style.cssText : flatCssText(style)).join('\n').trim();
  }
  return ''
}

function lightDomStyles(element: typeof ReactiveElement): void {
  const cssText = flatCssText(element.styles);
  if (cssText === '') return;
  const styleEl = document.createElement('style');
  styleEl.innerHTML = cssText;
  document.head.append(styleEl);
}
```

usage:
```ts
export class MyElement extends LitElement {
  static styles = css`
    my-element {
      color: green;
    }
  `;
}
lightDomStyles(MyElement);
```

**Comment by Moonnz** [comment]
The @mucsi96 way work well but isn't good for typing: 
```js
style.innerHTML = this.constructor.elementStyles; // TS2339: Property elementStyles does not exist on type Function
```
I prefer using this, and adding the style tag to the local component:
```js
style.innerHTML = flatCssText(MyElement.styles);
```

With the complete createRenderRoot function:
```js
    createRenderRoot() {
        if (this.querySelector(`style[data-tagname="${this.tagName}"]`)) {
            return this;
        }

        const style = document.createElement("style");
        style.innerHTML = flatCssText(MyElement.styles);
        style.setAttribute("data-tagname", this.tagName);
        this.append(style);
        return this;
    }
```

**Comment by justinfagnani** [comment]
That solution is going to append a `<style>` tag to every instance, which might be unnecessary.

You'd really want to append the tag to the instances root node, and only once.

**Comment by Totati** [comment]
Something like this?
``` javascript
  createRenderRoot() {
    const rootNode = this.getRootNode();
    const doc = 'head' in rootNode ? rootNode.head : rootNode;
    if (!doc.querySelector(`style[data-tagname="${this.tagName}"]`)) {
      const style = document.createElement("style");
      style.innerHTML = flatCssText(SimpleGreeting.styles);
      style.setAttribute("data-tagname", this.tagName);
      doc.append(style);
    }
    return this;
  }
  ```

**Comment by tomherni** [comment]
Stencil allows you to have web components without shadow DOM but still with style encapsulation. It's dirty, but it works _reasonably_ well. I'm currently on a project where we're migrating from Stencil to Lit. We have a few components that rely on this, which we just cannot change right now (due to enterprise & compatibility reasons...).

I wrote something, but can't share it, unfortunately. But you could take a look at [Stencil's source code](https://github.com/ionic-team/stencil/blob/a02bc3606f2e10331ff9aaa12618dbee271d86fa/src/runtime/styles.ts) and use that as a starting point. Essentially, you want to do the following:

* Find the appropriate container node for your component styles using `getRootNode()` (which would be either the document or an ancestor's shadow root).
* Check if the container node still requires your component styles by keeping track of container nodes & component names in a WeakMap (because you don't want multiple instances creating duplicate styles).
* If the container node does not have your component styles yet, then add them. If the container is a shadow root then add them to the adopted stylesheets (if supported), otherwise inject a `<style>` element.

I was able to skip some scenarios because I know _where_ and _how_ my components are consumed. But you may want to think about what happens to the injected component styles when your component is moved to another container, or when all instances are removed.

In the end, you do end up with quite a bit of logic to get everything working right, and it's not pretty since your component is making changes to its consumer's DOM. It makes sense that Lit doesn't cover this.

**Comment by itsmebasti** [comment]
Inspired by the previous examples, I came up with this. For me it is the most native solution I could come up with. I am curious if Justin has any improvements or feedback.

```ts
import { LitElement, html, css, adoptStyles } from 'lit-element';
import { customElement } from 'lit/decorators.js';

@customElement('my-component')
export class MyComponent extends LitElement {
  createRenderRoot() {
    adoptStyles(this.getRootNode() as ShadowRoot, [css`
      my-component div {
        color: green;
      }
    `]);
    
    return this;
  };
  
  override render() {
    return html`<div>hello lit</div>`
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'my-component': MyComponent;
  }
}
```
Notes: 
getRootNode() might return the document as well, but I typed it to ShadowRoot for adoptStyles() which seems to work for the document as well, so I am fine with the wrong type trade-off.

I had a previous versions adding the styles manually to prevent duplicates using `if(!rootNode.adoptedStyleSheets.includes(style))`, but found out I can further simplify everything after switching to adoptStyles, since after some testing it seemed to never create duplicates.

There is still one topic that, in my opinion, makes it non packageable for the lit core, which is the requirement to prepend `my-component` every css selector, to not break the hosting element. I guess if this would be solved internally it might imply too much hidden stuff on top of expected behavior which is probably not wanted by the lit owners.
